ta/pkcs11/src/pkcs11_token.c

```diff
@@ -26,13 +26,15 @@
 /* Static allocation of tokens runtime instances (reset to 0 at load) */
 struct ck_token ck_token[TOKEN_COUNT];
 
-/* Static allocation of tokens runtime instances */
 struct ck_token *get_token(unsigned int token_id)
 {
-	if (token_id > TOKEN_COUNT)
-		return NULL;
+	unsigned int id = 0;
+
+	for (id = 0; id < TOKEN_COUNT; id++)
```

Comment:
I'm not convinced this is speculation safe. It looks like it could easily be optimized away by the compiler.
Please use something from `<speculation_barrier.h>` instead.

Reply:
right, `id` could overrun `TOKEN_COUNT` in speculated execution. This bounds speculation execution to the depth of the speculation stack, but only reduces the attack surface.

`speculation_barrier.h` provides helper for loading operations, whereas here we only aim at safely validating the provided index.

Proposal, I hope not too complicated: have token instance address stored inside the structure, and use `load_no_speculate()` to load the targeted pointer inside range of `ck_token[]`.

Reply:
This is where `array_index_nospec()`  (`include/linux/nospec.h`)in the linux kernel would have been handy. I would really like to have a version of the with a permissive license.

Perhaps we can leave this open with just a comment about speculation right now.

Reply:
> with a permissive license

That's the point...
In the mean time, I propose https://github.com/OP-TEE/optee_os/pull/3648 to address the issue.

Reply:
After countless google attempts I finally accumulated enough karma to find this:
https://fuchsia.googlesource.com/fuchsia/+/master/zircon/system/ulib/fbl/include/fbl/confine_array_index.h

Reply:
Nice. Lacks 32bit support but looks addressable.
I'll create a P-R to get your feedback.

Reply:
https://github.com/OP-TEE/optee_os/pull/3649

Reply:
thanks, will use.